File: emcee_navigation/src/robot_driver.cpp
#include <iostream>

#include <ros/ros.h>
#include <geometry_msgs/Twist.h>
#include <tf/transform_listener.h>


// Reference: http://wiki.ros.org/pr2_controllers/Tutorials/Using%20the%20base%20controller%20with%20odometry%20and%20transform%20information
class RobotDriver
{
private:
  //! The node handle we'll be using
  ros::NodeHandle nh_;
  //! We will be publishing to the "cmd_vel" topic to issue commands
  ros::Publisher cmd_vel_pub_;
  //! We will be listening to TF transforms as well
  tf::TransformListener listener_;

public:
  //! ROS node initialization
  RobotDriver(ros::NodeHandle &nh)
  {
    nh_ = nh;
    //set up the publisher for the cmd_vel topic
    cmd_vel_pub_ = nh_.advertise<geometry_msgs::Twist>("/cmd_vel", 1);
}

  //! Drive forward a specified distance based on odometry information
  bool driveForwardOdom(double distance)
  {
    //wait for the listener to get the first message
    listener_.waitForTransform("/base_link", "/odom",
    ros::Time(0), ros::Duration(1000.0));

    //we will record transforms here
    tf::StampedTransform start_transform;
    tf::StampedTransform current_transform;

    //record the starting transform from the /odometry to the base frame
    listener_.lookupTransform("/base_link", "/odom",
    ros::Time(0), start_transform);

    //we will be sending commands of type "twist"
    geometry_msgs::Twist base_cmd;
    //the command will be to go forward at 0.25 m/s
    base_cmd.linear.y = base_cmd.angular.z = 0;
    base_cmd.linear.x = 0.25;

    ros::Rate rate(10.0);
    bool done = false;
    while (!done && nh_.ok())
    {
      cmd_vel_pub_.publish(base_cmd);
      rate.sleep();
      //get the current transform
      try
      {
        listener_.lookupTransform("/base_link", "/odom",
        ros::Time(0), current_transform);
      }
      catch (tf::TransformException ex)
      {
        ROS_ERROR("%s",ex.what());
        break;
      }
      //see how far we've traveled
      tf::Transform relative_transform = start_transform.inverse() * current_transform;
      double dist_moved = relative_transform.getOrigin().length();

      if(dist_moved > distance) done = true;
    }
    if (done) return true;
    return false;

  }
  bool turnOdom(bool clockwise, double radians)
  {
    while(radians < 0) radians += 2*M_PI;
    while(radians > 2*M_PI) radians -= 2*M_PI;

    //wait for the listener to get the first message
    listener_.waitForTransform("/base_link", "/odom",
    ros::Time(0), ros::Duration(1.0));

    //we will record transforms here
    tf::StampedTransform start_transform;
    tf::StampedTransform current_transform;

    //record the starting transform from the odometry to the base frame
    listener_.lookupTransform("/base_link", "/odom",
    ros::Time(0), start_transform);

    //we will be sending commands of type "twist"
    geometry_msgs::Twist base_cmd;
    //the command will be to turn at 0.75 rad/s
    base_cmd.linear.x = base_cmd.linear.y = 0.0;
    base_cmd.angular.z = 3.0;
    if (clockwise) base_cmd.angular.z = -base_cmd.angular.z;

    //the axis we want to be rotating by
    tf::Vector3 desired_turn_axis(0,0,1);
    if (!clockwise) desired_turn_axis = -desired_turn_axis;

    ros::Rate rate(10.0);
    bool done = false;
    while (!done && nh_.ok())
    {
      //send the drive command
      cmd_vel_pub_.publish(base_cmd);
      rate.sleep();
      //get the current transform
      try
      {
        listener_.lookupTransform("/base_link", "/odom",
        ros::Time(0), current_transform);
      }
      catch (tf::TransformException ex)
      {
        ROS_ERROR("%s",ex.what());
        break;
      }
      tf::Transform relative_transform =
      start_transform.inverse() * current_transform;
      tf::Vector3 actual_turn_axis =
      relative_transform.getRotation().getAxis();
      double angle_turned = relative_transform.getRotation().getAngle();
      if ( fabs(angle_turned) < 1.0e-2) continue;

      if ( actual_turn_axis.dot( desired_turn_axis ) < 0 )
      angle_turned = 2 * M_PI - angle_turned;

      if (angle_turned > radians) done = true;
    }
    if (done) return true;
    return false;
  }

};

int main(int argc, char** argv)
{
  //init the ROS node
  ros::init(argc, argv, "robot_driver");
  ros::NodeHandle nh;

  RobotDriver driver(nh);
  // driver.driveForwardOdom(0.5);
  driver.turnOdom(1, 1.570796);
  // ROS_INFO("Runs both");
  ros::Duration(1).sleep();
  driver.driveForwardOdom(4.0);
  ros::Duration(1).sleep();
  driver.turnOdom(1, 0.786398);
  ros::Duration(1).sleep();
  driver.driveForwardOdom(0.2);

  ROS_INFO("Navigation complete! :)");
}

File: emcee_navigation/scripts/robot-driver.py
#!/usr/bin/env python
import rospy
from geometry_msgs.msg import Twist
import tf
import math



class RobotDriver:
    def __init__(self):
        self.cmd_vel_pub = rospy.Publisher('/cmd_vel', Twist, queue_size=1)
        self.listener = tf.TransformListener()

    def drive_forward_odom(self, distance):
        start_transform = self.listener.lookupTransform('/base_link', '/odom', rospy.Time(0), timeout=rospy.Duration(100));
        self.listener.lookupTransform()

        base_cmd = Twist()
        base_cmd.linear.y = base_cmd.angular.z = 0;
        base_cmd.linear.x = 0.25

        rate = rospy.Rate(10)

        done = False
        while not done and not rospy.is_shutdown():
            self.cmd_vel_pub.publish(base_cmd)
            rate.sleep()

            try:
                current_transform = self.listener.lookupTransform('/base_link', '/odom', rospy.Time(0));
            except (tf.LookupException, tf.ConnectivityException, tf.ExtrapolationException) as e:
                rospy.loginfo(e)

            relative_transform = start_transform.inverse() * current_transform
            dist_moved = relative_transform.get_origin().length

def main():
    rospy.init_node('robot_driver')




def callback(objects):
    if (objects.data):
        obj_id = objects.data[0]
        rospy.loginfo(rospy.get_caller_id() + str(obj_id))
        print(obj_id)
        make_decisions(obj_id)



def make_decisions(obj_id):
    rate = rospy.Rate(10)

    rospy.loginfo(1)
    if (obj_id in odwallas):
        pub.publish(1)
    elif (obj_id in sprite):
        pub.publish(2)
    elif (obj_id in water):
        pub.publish(3)
    else:
        pub.publish(0)



def listener():
    rospy.Subscriber("objects", Float32MultiArray, callback)
    # spin() simply keeps python from exiting until this node is stopped
    rospy.spin()



if __name__ == '__main__':
    try:
        main();
    except rospy.ROSInterruptException:
        pass
